Kuadrant/kuadrant-operator
# fix: tlspolicy enforced condition when certificate/issuer ready condition is missing

# Description
Closes: https://github.com/Kuadrant/kuadrant-operator/issues/704

- fix: tlspolicy enforced condition when certificate/issuer ready condition is missing
  - Previous using IsStatusConditionFalse will return true if the Ready status condition is missing which caused TLSPolicy to incorrectly set the Enforced condition to true
- feat: TLSPolicy watch for owned Certificates
- feat: watch for issuer/clusterIssuer status changes for TLSPolicy
  - TLSPolicy should watch for status changes to ClusterIssuer/Issuer to react to whether the policy is correctly enforced or not. Also previously TLSPolicy integration tests are a bit flaky due to an certain order of events must happen / resources must be ready in order to pass the test
- regression: do not wait for gateways to be ready for tlspolicy
  - There is a regression issue where TLSPolicy should not wait for the gateways to be ready as the controller may be creating the neceesary resources in order for the gateway to be in the ready state
- fix: get all gateways events for tlspolicy
  - Previously the gateway event for TLSPolicy is filtered out if the gateway is not in a programmed state in the DAG, but is a valid use case for this policy. This caused the test `TLSPolicy controller with multiple https listener [It] should delete tls certificate when listener is removed` to be flaky. Added this filter option to the DAG to work as expected for TLSPolicy.

# Verification
Passing tests should be sufficient but if you want to manually verrify:

- Deploy this branch
```sh
make local-setup
```
- Apply `TLSPolicy`
```sh
kubectl apply -n istio-system -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: istio-ingressgateway
spec:
  gatewayClassName: istio
  listeners:
  - name: http
    hostname: '*.toys.io'
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - name: toys
        kind: Secret
    allowedRoutes:
      namespaces:
        from: All    
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-issuer
spec:
  selfSigned: {}
---
apiVersion: kuadrant.io/v1alpha1
kind: TLSPolicy
metadata:
  name: gw-tls
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: istio-ingressgateway
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: selfsigned-issuer
EOF
```
- Watch TLSPolicy status
```sh
watch -n 0.1 "kubectl get tlspolicy -A -o yaml | yq '.items[0]'"
```
- Verified enforced status is:
```yaml
    - lastTransitionTime: "2024-07-01T08:59:46Z"
      message: TLSPolicy has been successfully enforced
      reason: Enforced
      status: "True"
      type: Enforced
```
- Delete certificate secret
```sh
kubectl delete secrets toys -n istio-system
```
- Verify `TLSPolicy` enforced condition quickly tranistions to `false` and back to `true`
